File: ruby/coding_challenge.rb
require 'pry'

class StateCapitals
  attr_reader :states,
              :capitals

  def initialize
    @states = {"Oregon" => "OR",
              "Alabama" => "AL",
              "New Jersey" => "NJ",
              "Colorado" => "CO"}
    @capitals = {"OR" => "Salem",
                "AL" => "Montgomery",
                "NJ" => "Trenton",
                "CO" => "Denver"}
  end

  def find_abbreviation(state)
    states[state]
  end

  def find_capital(abbrev)
    capitals[abbrev]
  end

  def find_capital_from_state(state)
    if capitals[find_abbreviation(state)]
      capitals[find_abbreviation(state)]
    else
      "Unknown"
    end
  end

  def capital_to_abbreviation(city)
    capitals.key(city)
  end

  def capital_to_state(city)
    states.key(capital_to_abbreviation(city))
  end
end

File: javascript/codingChallenge.js
const _ = require('lodash')

class StateCapitals {
  constructor() {
    this.states = {"Oregon": "OR",
              "Alabama": "AL",
              "New Jersey": "NJ",
              "Colorado": "CO"};

    this.capitals = {"OR": "Salem",
                "AL": "Montgomery",
                "NJ": "Trenton",
                "CO": "Denver"};
  }

  state_to_abbreviation(state) {
    return this.states[state]
  }

  state_to_capital(state) {
    if (this.capitals[this.state_to_abbreviation(state)]) {
      return this.capitals[this.state_to_abbreviation(state)]
    } else {
      return "Unknown"
    }
  }

  capital_to_abbreviation(city) {
    return _.findKey(this.capitals, _.partial(_.isEqual,city))
  }
}

module.exports = StateCapitals;
